doc/compiler/il/MainlineAndOutOfLineCode.md

```diff
@@ -0,0 +1,47 @@
+<!--
+Copyright (c) 2021, 2021 IBM Corp. and others
+
+This program and the accompanying materials are made available under
+the terms of the Eclipse Public License 2.0 which accompanies this
+distribution and is available at https://www.eclipse.org/legal/epl-2.0/
+or the Apache License, Version 2.0 which accompanies this distribution and
+is available at https://www.apache.org/licenses/LICENSE-2.0.
+
+This Source Code may also be made available under the following
+Secondary Licenses when the conditions for such availability set
+forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+General Public License, version 2 with the GNU Classpath
+Exception [1] and GNU General Public License, version 2 with the
+OpenJDK Assembly Exception [2].
+
+[1] https://www.gnu.org/software/classpath/license.html
+[2] http://openjdk.java.net/legal/assembly-exception.html
+
+SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+-->
+
+# Mainline and Out of Line Code
+The operations that are believed to run commonly are arranged on the mainline path. The less often executed operations are chosen as the branch target on the out of line (`OOL`) path. Arranging the code that executes more often in the mainline takes advantage of branch prediction and instruction cache resources. The out of line path is usually in a different cache line for an infrequently executed operation. By falling through on the mainline path, the work could be done in the same cache line, which avoids jumping to another spot to bring in an entire new cache line to the cache. Occasionally, there would be some minor branching within the code on the mainline path to implement some basic logics.
+
+Executing the common code on the mainline is not only adopted in IL but also in the codegen, such as the non-trivial logics: write barrier, instanceof, etc. One example in the codegen is the bound check. It does the comparison between arraylength and index. If it fails, an exception is going to be raised and it branches to an `OOL`, or a snippet that jumps into the VM to throw the exception that needs to the thrown. Because the path that jumps to throw the exception is expected to be very rarely called, the snippet that calls into the VM to throw the appropriate exception is put at the end of the code section, or `OOL`, at the time of generating code for the bound check, so that it does not interfere with the code that we believe will actually run. It is possible that subsequent bound checks (or other checks/operations) that get evaluated also append more code to the end of the code section. The snippet for a given bound check would not be exactly at the very end but instead, near the end of the code section.
```

Comment:
> non-trivial logics: write barrier, instanceof, etc.

These examples seem more fitting to OpenJ9. Perhaps there are other examples within OMR we can use.

Reply:
I'm not really sure. I thought they were applicable to OMR as well. Any suggestion?  @vijaysun-omr is there any example that is more applicable to OMR?

Reply:
`iwrtbar`, `new` and `instanceof` all seem to be opcodes present in OMR, so I'm unsure what you might be saying that makes it unsuited for OMR documentation. All that you seem to be saying is that the logic may be non-trivial it seems, and that is probably the case regardless of the language being spoken about. 

Other operations with some internal control flow could be 64-bit operations (such as long comparisons) on 32-bit platforms.